### PR TITLE
Null out `ProxyConfiguration.userName` on save

### PR DESCRIPTION
**Pull request by jglick**
Analysis of a bug report for CloudBees CI showed that when a proxy was configured with a username & password, uses of `newHttpClientBuilder` would (silently!) fail to pass an `Authorization` header to the real server, but only on Java 17 or newer. (As of https://github.com/openjdk/jdk/commit/028f2e14b3550dd1f2a358a2a0aaf83d4a824af0 I think; the same code works on Java 11.) This issue is described in [JDK-8306745](https://bugs.openjdk.org/browse/JDK-8306745) complete with a unit test, but alas that was closed. The test there looks similar to the Jenkins authenticator, including the logic https://github.com/jenkinsci/jenkins/blob/f7fea8d891dccedd6b16391ade1bef95374e8207/core/src/main/java/hudson/ProxyConfiguration.java#L162 which correctly reserves the authentication for the proxy server. So far as I can tell there is no practical way to make `newHttpClientBuilder` behave correctly on Java 17 when an authenticated proxy is configured: the bug evaluation says you should

> [not] use any authenticator, and explicitly handle 401/407 in the calling code

which is only an option if the complete usage including request is redesigned to manually check for a 407 and retry with a `Proxy-Authorization` header, but that logic cannot be encapsulated in a `HttpClient.Builder` return value unless Jenkins implemented its own `Builder` and a special delegating implementation of `HttpClient`.

While looking at that, I saw that the reproducer claimed the bug occurred even when a username was _not_ set, which seems to be because the GUI configuration of the proxy after #3935 saved `""` rather than `null` when the username text field was left blank. This at least is easy to fix, limiting the scope of the problem.

### Testing done

Without `setUserName` change, test fails

```
java.lang.AssertionError: expected null, but was:<>
	at org.junit.Assert.fail(Assert.java:89)
	at org.junit.Assert.failNotNull(Assert.java:756)
	at org.junit.Assert.assertNull(Assert.java:738)
	at org.junit.Assert.assertNull(Assert.java:748)
	at hudson.ProxyConfigurationManagerGUITest.lambda$configRoundtrip$1(ProxyConfigurationManagerGUITest.java:51)
```

### Proposed changelog entries

- Proxy configuration saved via the GUI always configured an authenticator even if the username was blank.

Before the changes are marked as `ready-for-merge`:

```[tasklist]
### Maintainer checklist
- [ ] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [ ] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [ ] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [ ] Proper changelog labels are set so that the changelog can be generated automatically.
- [ ] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).
```

### CODE REVIEW

File: core/src/main/java/hudson/ProxyConfiguration.java
Comment (Vlatombe) [inline core/src/main/java/hudson/ProxyConfiguration.java:171]: Why not handle it with `readResolve`?

**Review by Vlatombe — APPROVED** [review]

**Review by Vlatombe — APPROVED** [review]

**Review by res0nance — APPROVED** [review]

**Review by timja — APPROVED** [review]

**Review by StefanSpieker — APPROVED** [review]

**Comment by NotMyFault** [comment]
/label ready-for-merge

---

This PR is now ready for merge. We will merge it after ~24 hours if there is no negative feedback.
Please see the [merge process documentation](https://github.com/jenkinsci/jenkins/blob/master/docs/MAINTAINERS.adoc#merge-process) for more information about the merge process.
Thanks!

**Pull request merged**
